contracts/src/v0.8/functions/dev/1_0_0/FunctionsSubscriptions.sol

```diff
@@ -0,0 +1,560 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+
+import {IFunctionsSubscriptions} from "./interfaces/IFunctionsSubscriptions.sol";
+import {ERC677ReceiverInterface} from "../../../interfaces/ERC677ReceiverInterface.sol";
+import {LinkTokenInterface} from "../../../interfaces/LinkTokenInterface.sol";
+import {IFunctionsBilling} from "./interfaces/IFunctionsBilling.sol";
+import {IFunctionsRouter} from "./interfaces/IFunctionsRouter.sol";
+import {SafeCast} from "../../../shared/vendor/openzeppelin-solidity/v.4.8.0/contracts/utils/SafeCast.sol";
+
+/**
+ * @title Functions Subscriptions contract
+ * @notice Contract that coordinates payment from users to the nodes of the Decentralized Oracle Network (DON).
+ * @dev THIS CONTRACT HAS NOT GONE THROUGH ANY SECURITY REVIEW. DO NOT USE IN PROD.
+ */
+abstract contract FunctionsSubscriptions is IFunctionsSubscriptions, ERC677ReceiverInterface {
+  // Reentrancy protection.
+  bool internal s_reentrancyLock;
+  error Reentrant();
+
+  LinkTokenInterface private LINK;
+
+  // ================================================================
+  // |                      Subscription state                      |
+  // ================================================================
+
+  // We make the sub count public so that its possible to
+  // get all the current subscriptions via getSubscription.
+  uint64 private s_currentsubscriptionId;
+
+  // s_totalBalance tracks the total LINK sent to/from
+  // this contract through onTokenTransfer, cancelSubscription and oracleWithdraw.
+  // A discrepancy with this contract's LINK balance indicates that someone
+  // sent tokens using transfer and so we may need to use recoverFunds.
+  uint96 private s_totalBalance;
+
+  mapping(uint64 => IFunctionsSubscriptions.Subscription) /* subscriptionId */ /* subscription */
+    private s_subscriptions;
+
+  // We need to maintain a list of addresses that can consume a subscription.
+  // This bound ensures we are able to loop over them as needed.
+  // Should a user require more consumers, they can use multiple subscriptions.
+  uint16 public constant MAX_CONSUMERS = 100;
+  mapping(address => mapping(uint64 => IFunctionsSubscriptions.Consumer)) /* consumer */ /* subscriptionId */ /* Consumer data */
+    private s_consumers;
+
+  event SubscriptionCreated(uint64 indexed subscriptionId, address owner);
+  event SubscriptionFunded(uint64 indexed subscriptionId, uint256 oldBalance, uint256 newBalance);
+  event SubscriptionConsumerAdded(uint64 indexed subscriptionId, address consumer);
+  event SubscriptionConsumerRemoved(uint64 indexed subscriptionId, address consumer);
+  event SubscriptionCanceled(uint64 indexed subscriptionId, address to, uint256 amount);
```

Comment:
For clarify, let's rename `to` and `amount` to something like `refundAddress`, `refundAmount`.